### DISCUSSION TITLE
About :setlocal backspace=nostop

### DISCUSSION THREAD

**Discussion by michelesr** [hrsh7th/nvim-cmp#2116] [General]
https://github.com/hrsh7th/nvim-cmp/blob/main/lua/cmp/utils/feedkeys.lua#L17 uses nvim feedkeys to set the `backspace=nostop` option temporarily. That option is global, so `setlocal` has the same effect of `set`. In addition to that, using feedkeys to set it makes it hard to understand it was set from a script (e.g. with `:verbose set bs?`). I have a problem where sometimes `backspace` gets set to `nostop`, and while I'm not sure it's nvim-cmp doing that and failing to restore the previous value after, it's hard to understand what's effectively setting that and why.

**Comment by michelesr** [comment]
My question is, if effectively those options needs to be set in order to feed the desired keys effectively, can they be set programmatically rather than by feeding characters to nvim (e.g. `nvim.go.backspace = 'nostop'`)? That way if an option gets stuck somehow it's easier to understand that it was set by nvim-cmp.

**Reply by Shougo** [reply]
Yes, it should be set instead.

**Comment by michelesr** [comment]
I keep getting `bs` set randomly to `nostop` and I have to restore it by manually running `:set bs=indent,eol,start`, and looking through my plugins nvim-cmp is the only one setting that option. Since it happens randomly I have no idea how to reproduce. Did someone experience a similar problem or can suggest some reproduction steps?

**Reply by michelesr** [reply]
```vim
augroup BackspaceOptionChange
  autocmd!
  autocmd OptionSet backspace echom "backspace option changed to: " . &backspace
augroup END
```

This can be used to track changes to the option. By accepting a completion in nvim the change triggered and the option was never restored to the original value. This happens sporadically and didn't find reproduction steps that are reliable unfortunately :-(

**Reply by michelesr** [reply]
Ended up using this:

```lua
local feedkeys = require('cmp.utils.feedkeys')
-- patch the feedkeys function to prevent unwanted changes to the bs option
local fk_mt = getmetatable(feedkeys.call)
local orig_f = fk_mt.__call
local function patched_feedkeys(self, keys, mode, callback)
  local bs = vim.go.backspace
  local _, err = pcall(function()
    orig_f(self, keys, mode, callback)
  end)
  if err then
    vim.notify('nvim-cmp feedkeys error: ' .. err)
  end
  if vim.go.backspace ~= bs then
    print('Detected undesired value of backspace:  ' .. vim.go.backspace)
    print('Restoring backspace option to ' .. bs)
    vim.go.backspace = bs
  end
end
fk_mt.__call = patched_feedkeys
```

I'm not entirely sure it works but so far I didn't get problems after applying this.

**Comment by michelesr** [comment]
I think the situation where nostop is not restored because an error is raised before it can do so can be ruled out because the error would appear in messages, and nvim_feedkeys() seems to batch keystrokes before actually executing them so if there was an error in the nvim-cmp feedkeys cmp (for example invalid input) that it's likely to fail before the backspace option can be set.

Could there be a race condition where two instances of feedkeys are run concurrently and the 2nd catches the value of backspace to be nostop and then restores that value at the end of execution after the 1st one already restored the original one? Seems far fetched, but you never know.

**Comment by michelesr** [comment]
I assume nobody else had this issue, but just in case, I'll share my [config](https://github.com/michelesr/nvim-config/blob/250b8185a56bd1ac68a6f750bb204e4c7ec0d695/lua/config/cmp.lua#L13-L48) that wraps the feedkeys function so that it saves the values of backspace and lazyredraw options and sets an autocmd to automatically restore them when all the inputs have been handled. It also prints a debug message when this happens (it happens rarely, but it does indeed happen).
